### PR TITLE
feat: add metada to base strategy

### LINKED ISSUE #165: Add Metadata URI to BaseStrategy/Strategy
add a new public storage variable `metadataURI`, which can be edited by the strategist and governance, and contains a URI link to some publicly accessible CDN (such as IPFS e.g. `ipfs://Qm....`).

This link should be content-addressed if possible to avoid rewriting the content without an on-chain audit history.

The content itself should be JSON encoded, and list things such as a description of the Strategy, what tokens it involves, failure modes, etc.

Also consider replacing `Strategy.name()` with this field

### PR DESCRIPTION
**Pull request by steffenix**
closes #165 

I have added the metadataURI to the constructor, I am not sure if that is a requirement ?

### CODE REVIEW

File: contracts/BaseStrategy.sol [hunk 1]
```diff
@@ -228,14 +231,16 @@ abstract contract BaseStrategy {
      *  contract is deployed.
      * @dev `_vault` should implement `VaultAPI`.
      * @param _vault The address of the Vault responsible for this Strategy.
+     * @param _metadataURI The URI that describe the strategy.
      */
-    constructor(address _vault) public {
+    constructor(address _vault, string memory _metadataURI) public {
```
Comment (fubuloubu) [inline contracts/BaseStrategy.sol:236]: Don't add it to the constructor

**Review by fubuloubu — APPROVED** [review]

File: tests/functional/strategy/test_misc.py [hunk 2]
```diff
@@ -101,3 +101,10 @@ def test_reject_ether(gov, strategy):
     # Fallback fails too
     with brownie.reverts("Cannot send ether to nonpayable function"):
         gov.transfer(strategy, 1)
+
+
+def test_set_metadataURI(gov, strategy, rando):
+    strategy.setMetadataURI("ipfs://test", {"from": gov})
+    strategy.setMetadataURI("ipfs://test", {"from": gov})
```
Comment (fubuloubu) [inline tests/functional/strategy/test_misc.py:108]:
```suggestion
    assert strategy.metadataURL() == ""  # Empty by default
    strategy.setMetadataURI("ipfs://test", {"from": gov})
    assert strategy.metadataURL() == "ipfs://test"
    strategy.setMetadataURI("ipfs://test2", {"from": gov})
    assert strategy.metadataURL() == "ipfs://test2"
```

**Review by fubuloubu — CHANGES_REQUESTED** [review]

**Comment by fubuloubu** [comment]
Address above comments and good to merge!

File: contracts/BaseStrategy.sol [hunk 3]
```diff
@@ -186,6 +187,8 @@ abstract contract BaseStrategy {
 
     event EmergencyExitEnabled();
 
+    event UpdatedMetaDataUri(string metadataURI);
```
Comment (fubuloubu) [inline contracts/BaseStrategy.sol:190]:
```suggestion
    event UpdatedMetadataURI(string metadataURI);
```

File: contracts/BaseStrategy.sol [hunk 4]
```diff
@@ -335,6 +340,19 @@ abstract contract BaseStrategy {
...
+     * @notice
+     *  Used to change `metadataURI`. `metadataURI` is used to store the URI
+     * of the file describing the strategy.
+     *
+     *  This may only be called by governance or the strategist.
+     * @param _metadataURI The URI that describe the strategy.
+     */
+    function setMetadataURI(string calldata _metadataURI) external onlyAuthorized {
+        metadataURI = _metadataURI;
+        emit UpdatedMetaDataUri(_metadataURI);
```
Comment (fubuloubu) [inline contracts/BaseStrategy.sol:353]:
```suggestion
        emit UpdatedMetadataURI(_metadataURI);
```

**Review by fubuloubu — CHANGES_REQUESTED** [review]
nit: because this affects the event id

File: contracts/test/TestStrategy.sol [hunk 5]
```diff
@@ -12,7 +12,7 @@ import {BaseStrategy, StrategyParams, VaultAPI} from "../BaseStrategy.sol";
 contract TestStrategy is BaseStrategy {
     bool public doReentrancy;
 
-    constructor(address _vault) public BaseStrategy(_vault) {}
+    constructor(address _vault) public BaseStrategy(_vault, "") {}
```
Comment (fubuloubu) [inline contracts/test/TestStrategy.sol:15]: revert this change back

**Review by fubuloubu — APPROVED** [review]
Fix up your commit message if you can, so it passes commit lint

**Comment by steffenix** [comment]
> Fix up your commit message if you can, so it passes commit lint

I did made a mistake with the suggested changes commit message. It's fixed.

**Comment by steffenix** [comment]
ready

**Review by dougstorm — APPROVED** [review]

**Pull request merged**